### PR TITLE
Support #[global_allocator] without the allocator shim

### PR DESCRIPTION
**Pull request by bjorn3**
This makes it possible to use liballoc/libstd in combination with `--emit obj` if you use `#[global_allocator]`. This is what rust-for-linux uses right now and systemd may use in the future. Currently they have to depend on the exact implementation of the allocator shim to create one themself as `--emit obj` doesn't create an allocator shim.

Note that currently the allocator shim also defines the oom error handler, which is normally required too. Once `#![feature(default_alloc_error_handler)]` becomes the only option, this can be avoided. In addition when using only fallible allocator methods and either `--cfg no_global_oom_handling` for liballoc (like rust-for-linux) or `--gc-sections` no references to the oom error handler will exist.

To avoid this feature being insta-stable, you will have to define `__rust_no_alloc_shim_is_unstable` to avoid linker errors.

(Labeling this with both T-compiler and T-lang as it originally involved both an implementation detail and had an insta-stable user facing change. As noted above, the `__rust_no_alloc_shim_is_unstable` symbol requirement should prevent unintended dependence on this unstable feature.)

### CODE REVIEW

**Comment by jyn514** [comment]
Hmm, r? @scottmcm maybe?

File: compiler/rustc_codegen_cranelift/src/allocator.rs [hunk 1]
```diff
@@ -31,71 +32,75 @@ fn codegen_inner(
...
+                }
+            };
+
+            let sig = Signature {
+                call_conv: CallConv::triple_default(module.isa().triple()),
+                params: arg_tys.iter().cloned().map(AbiParam::new).collect(),
+                returns: output.into_iter().map(AbiParam::new).collect(),
+            };
+
+            let caller_name = format!("__rust_{}", method.name);
```
Comment (RalfJung) [inline compiler/rustc_codegen_cranelift/src/allocator.rs:69]: Should this use `global_fn_name`?

Comment (bjorn3) [reply compiler/rustc_codegen_cranelift/src/allocator.rs]: Done

**Comment by nbdd0121** [comment]
It took me a few while to understand this, so essentially instead of generating `__rust_alloc` that forwards to `__rg_alloc` or `__rdl_alloc` based on allocator kind, this PR makes `#[global_allocator]` generate `__rust_alloc` directly, while in case global allocator is absent, generate a shim that forwards `__rust_alloc` to `__rdl_alloc`.

Would it make sense to just generate a
```rust
#[global_allocator]
static GLOBAL: System = System;
```
at somewhere in higher level instead of having the logic duplicated in codegen? This would allow `__rdl_alloc` etc to be removed completely (and is more consistent with what document describes in https://doc.rust-lang.org/std/alloc/struct.System.html).

**Comment by bjorn3** [comment]
> Would it make sense to just generate a
> 
> ```rust
> #[global_allocator]
> static GLOBAL: System = System;
> ```
> 
> at somewhere in higher level instead of having the logic duplicated in codegen? This would allow `__rdl_alloc` etc to be removed completely (and is more consistent with what document describes in https://doc.rust-lang.org/std/alloc/struct.System.html).

You can use say `--crate-type cdylib --crate-type lib`, in which case the `cdylib` would need the allocator shim, but the `lib` must not get the allocator shim. Both use the same object files, except for the allocator shim object file that only the `bin` gets.

File: compiler/rustc_builtin_macros/src/global_allocator.rs [hunk 2]
```diff
@@ -82,14 +81,24 @@ impl AllocFnFactory<'_, '_> {
         let result = self.call_allocator(method.name, args);
         let (output_ty, output_expr) = self.ret_ty(&method.output, result);
         let decl = self.cx.fn_decl(abi_args, ast::FnRetTy::Ty(output_ty));
-        let header = FnHeader { unsafety: Unsafe::Yes(self.span), ..FnHeader::default() };
+        let header = FnHeader {
+            unsafety: Unsafe::Yes(self.span),
+            ext: ast::Extern::from_abi(Some(ast::StrLit {
```
Comment (nbdd0121) [inline compiler/rustc_builtin_macros/src/global_allocator.rs:86]: `alloc/src/alloc.rs` uses `extern "Rust"` for `__rust_alloc` and friends. Seems like a mismatch?

Comment (bjorn3) [reply compiler/rustc_builtin_macros/src/global_allocator.rs:86]: Indeed. I thought it used `extern "C"`. I will need to revert this change.

**Comment by JohnCSimon** [comment]
merge conflicts

**Comment by bjorn3** [comment]
Rebased

**Comment by nbdd0121** [comment]
The ABI is wrong again

**Comment by bjorn3** [comment]
Fixed. I think I authored the ABI revert on a different device and didn't pull it before rebasing.

**Comment by scottmcm** [comment]
Sorry for missing this, but I don't think I can't provide useful guidance here.

r? compiler

**Comment by bjorn3** [comment]
Fixing https://github.com/rust-lang/rust/issues/95113 may require applying this approach to alloc error handler too.

File: compiler/rustc_codegen_llvm/src/allocator.rs [hunk 3]
```diff
@@ -114,8 +118,7 @@ pub(crate) unsafe fn codegen(
         attributes::emit_uwtable(llfn, true);
     }
 
-    let kind = if has_alloc_error_handler { AllocatorKind::Global } else { AllocatorKind::Default };
-    let callee = kind.fn_name(sym::oom);
+    let callee = if has_alloc_error_handler { "__rg_oom" } else { "__rdl_oom" };
```
Comment (pnkfelix) [inline compiler/rustc_codegen_llvm/src/allocator.rs:121]: nit: it seems bad to embed the knowledge of what prefix to use directly into the code. Elsewhere, you have helper functions that add the appropriate prefix to use; why not apply that pattern here? Is it because it wouldn't get used elsewhere in the compiler?

Comment (bjorn3) [reply compiler/rustc_codegen_llvm/src/allocator.rs:121]: I don't quite recall. I will take a look tomorrow and maybe revert this.

File: compiler/rustc_ast/src/expand/allocator.rs [hunk 4]
```diff
@@ -1,18 +1,17 @@
...
 
-impl AllocatorKind {
-    pub fn fn_name(&self, base: Symbol) -> String {
-        match *self {
-            AllocatorKind::Global => format!("__rg_{}", base),
-            AllocatorKind::Default => format!("__rdl_{}", base),
-        }
-    }
+pub fn global_fn_name(base: Symbol) -> String {
+    format!("__rust_{}", base)
```
Comment (pnkfelix) [inline compiler/rustc_ast/src/expand/allocator.rs:10]: Can you also update the comments elsewhere in the compiler+stdlib code to reflect the new structure being deployed here?

Here are the references my naive grep yielded:

https://github.com/rust-lang/rust/blob/99da9aec246e6b500a4e589e4b9096306bb89b07/library/alloc/src/alloc.rs#L23-L27

https://github.com/rust-lang/rust/blob/99da9aec246e6b500a4e589e4b9096306bb89b07/library/alloc/src/alloc.rs#L354-L356

https://github.com/rust-lang/rust/blob/297a8018b525c28ef10ee6a91d61954839b508b9/compiler/rustc_monomorphize/src/partitioning/default.rs#L517-L521

**Comment by pnkfelix** [comment]
What happens to current uses of `_rg_`-prefixed functions that I have found in Miri and Cranelift?

e.g.

https://github.com/rust-lang/miri/blob/c1bbf078f368b1a16951730148fe15127df249ef/src/shims/foreign_items.rs#L44 (among others)

and

https://github.com/rust-lang/rust/blob/297a8018b525c28ef10ee6a91d61954839b508b9/compiler/rustc_codegen_cranelift/src/allocator.rs#L111

**Comment by bjorn3** [comment]
> What happens to current uses of _rg_-prefixed functions that I have found in Miri and Cranelift?

They will need to be updated.

**Comment by pnkfelix** [comment]
Overall, I think the problem described here is worth solving.

My main reaction upon reading this is "Surely it cannot be this simple?"

In other words: I'm asking myself: What features of the old system, with the level of indirection between the `_rust_`-prefixed functions that make up the shim, and the `_rg_`-prefixed functions that the shim methods called into .... what problem was *that* solving? Why can you discard all that stuff now?

Is the overarching lesson here that the shim was there first, back before we had `#[global_allocator]`, and then when `#[global_allocator]` was added, someone decided to reuse the shim-machinery, but such reuse didn't provide any value to Rust users (i.e., that if we had thought about it more carefully, we could have done `#[global_allocator]` this way from the start)?

**Comment by pnkfelix** [comment]
> Is the overarching lesson here that the shim was there first, back before we had `#[global_allocator]`, and then when `#[global_allocator]` was added, someone decided to reuse the shim-machinery, but such reuse didn't provide any value to Rust users (i.e., that if we had thought about it more carefully, we could have done `#[global_allocator]` this way from the start)?

Assuming the answer to this question is "yes, we could have always done it this way", then I think we should move ahead with this change.

From what I understand, the change to the symbol names being emitted is 100% implementation detail; developers were not able to link to the `_rg_`-prefixed symbols, so that aspect of this change is safe.

The only real problem is the one that @bjorn3 identified from the start: this is providing insta-stable functionality.

**Comment by bjorn3** [comment]
> Is the overarching lesson here that the shim was there first, back before we had #[global_allocator], and then when #[global_allocator] was added, someone decided to reuse the shim-machinery, but such reuse didn't provide any value to Rust users (i.e., that if we had thought about it more carefully, we could have done #[global_allocator] this way from the start)?

I think so. Would have to look at the commit history to know for sure though.

File: compiler/rustc_codegen_llvm/src/allocator.rs (hunk 3 above)
Comment (bjorn3) [reply compiler/rustc_codegen_llvm/src/allocator.rs]: Added a helper function. The existing `AllocatorKind::fn_name` helper was changed to use `__rust_*` for AllocatorKind::Global, but this still needs `__rg_oom`.

**Comment by bjorn3** [comment]
Rebased but not yet ready for review.

**Comment by JohnCSimon** [comment]
>  Rebased but not yet ready for review.


@rustbot author

**Comment by pnkfelix** [comment]
Discussed in T-compiler meeting. @bjorn3 says they still have more work to do on this

@rustbot label: +S-waiting-on-author -S-waiting-on-review

**Comment by bjorn3** [comment]
Rebased and made it non-insta stable as requested at https://rust-lang.zulipchat.com/#narrow/stream/197181-t-libs.2Fwg-allocators/topic/Support.20.23.5Bglobal_allocator.5D.20without.20the.20allocator.20shim.20.238684/near/277476902 It could result in a slight perf decrease due to an extra load being inserted in the alloc function to ensure that `__rust_no_alloc_shim_is_unstable` exists. As such I'm going to do a perf run.

@bors try @rust-timer queue

**Comment by wesleywiser** [comment]
Nominating for T-lang discussion. This came up during the compiler team triage meeting and we noticed that there hasn't been any official comment from T-lang on this yet. This PR is nearly ready to merge (a review is outstanding) and @bjorn3 noted there is desire to stabilize this eventually so your input would be appreciated!

**Comment by mhammerly** [comment]
> As side effect this will make #[global_allocator] usable on some platforms when libstd is dynamically linked. This does require the #[global_allocator] to be either in the executable or a dylib that is loaded before the dylib that has libstd linked into it.

As-is this PR fits well with https://github.com/rust-lang/rust/issues/73632 and it'd be valuable to merge, but I think before stabilizing anything we should explore stronger support for dylibs in general to make sure we don't design ourselves into a corner.

The "no `#[no_std]` in dylibs" problem has become a big one for me so I was going to ramp up on the area to see if I can come up with anything. I'll be sure to share if I do!

**Comment by bjorn3** [comment]
You can combine `#![no_std]` with dylibs just fine. It is the combination of a dynamically linked libstd and `#[global_allocator]` that doesn't work. I don't think there exists a portable solution for that. Even ignoring weak linkage not being portable enough, there is another problem. There may at most be 1 `#[global_allocator]` in the whole crate graph. There is no way to guarantee this once dylibs come into play.

**Comment by mhammerly** [comment]
> You can combine #![no_std] with dylibs just fine

I don't want to sidetrack this PR but this breaks down in mixed-language projects. Say you have a C++ binary with three mixed-language shared libraries A, B, and C. A and B each share a dep on, say, `unicode-ident`, while B and C shared a dep on `bitflags`. Each should be built as a shared library, but to do so each must provide a panic handler / eh_personality, of which there may only be one between them. Also, `core` cannot be built as a `dylib` so both `unicode-ident` and `bitflags` will statically link `core` which is a no-go. And `cdylib`, like `staticlib`, is not desirable in projects like this.

I'll look for or create a more relevant issue for that discussion, I just brought it up as the reason why I'll be ramping up on the space.

> There may at most be 1 #[global_allocator] in the whole crate graph. There is no way to guarantee this once dylibs come into play.

There's no way for `rustc` to guarantee it in a large mixed-language project even with static linkage. One project may have multiple separate "crate graphs" in its overarching dependency tree but may only have one `#[global_allocator]` between them.

I think Rust should start designing with the mixed-language Buck/Bazel use case in mind, not just Rust-centric Cargo. Not saying Rust should fully pivot, but from our perspective some of these guardrails and opinionated behaviors are not really important or are themselves problems.

> It is the combination of a dynamically linked libstd and #[global_allocator] that doesn't work. I don't think there exists a portable solution for that

What if we moved to using explicit sysroot dependencies ([wg-std-aware-cargo](https://github.com/rust-lang/wg-cargo-std-aware)) and the build system (Cargo, Buck, what have you) allowed you to configure a `#[global_allocator]` (or `#[panic_handler]`) to be supplied as a dep of `std`?

That specific idea may not be the solution (it might require `#[global_allocator]`s not to dep on `std` which I think is a new restriction?) but I'm sure we can think of something if we declare it a requirement for stabilization.

**Comment by nbdd0121** [comment]
I don't think that's relevant. If you have multiple separate crate graphs the allocator shims are already emitted multiple times and you will have symbol conflicts.

**Comment by bjorn3** [comment]
> There's no way for rustc to guarantee it in a large mixed-language project even with static linkage. One project may have multiple separate "crate graphs" in its overarching dependency tree but may only have one #[global_allocator] between them.

If within a single crate graph there is more than one global allocator, rustc will error. Between multiple crate graphs (which each get their own copy of libcore and any other crate) there is no issue as you can't pass rust types between them anyway and the `#[global_allocator]` won't try to override each other. `#[global_allocator]` and `#[panic_handler]` are local to a single crate graph.

> (it might require #[global_allocator]s not to dep on std which I think is a new restriction?)

That restriction will be a breaking change that makes `#[global_allocator]` unusable for many use cases.

**Comment by mhammerly** [comment]
To be clear, my request is that we think through options for dylibs before merging this PR in case it paints us into a corner. If we don't think it does, merge away.

**Comment by mhammerly** [comment]
> I don't think that's relevant. If you have multiple separate crate graphs the allocator shims are already emitted multiple times and you will have symbol conflicts.

Read through #73632; we are discussing stabilizing rlib's current behavior to help mixed-language projects avoid diamond dependencies from multiple `staticlib`s. If we do that, no allocator shims are generated at all for rlib-only projects. This PR would provide an officially-supported way for rlib-only projects to get allocator symbols when using static linkage because it skips the shim and just generates the allocator symbols, and it'll be on developers to ensure only one is provided. I mention it because from our perspective Rust already can't enforce that there's only one `#[global_allocator]` so that's not a counterpoint against supporting dylibs.

Agree on moving further dylib discussion to Zulip or a dedicated issue, sorry for the noise here.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting at length.

This in in the broad category of linking-related and system-related issues that has some overlap between compiler and lang, and it's not completely clear where all the jurisdictions lie. However, we did come to one clear consensus that we felt was entirely in lang's purview:

We don't think, at this time, that we want to constrain all future design space by committing to the *default* output format (rlib) being something that works with any linker. We think it would make sense to experiment with *opting in* to a new format that's designed to be linked by a different linker. Such an opt in would ensure that in the future, if we want to add a feature that we don't know how to support with a different linker, we can detect the conflict between that feature and the use of the third-party-linking-compatible format.

**Comment by bjorn3** [comment]
> We don't think, at this time, that we want to constrain all future design space by committing to the default output format (rlib) being something that works with any linker.

The current PR still keeps this unstable by requiring a static to be provided whose existence and name is unstable.

> We think it would make sense to experiment with opting in to a new format that's designed to be linked by a different linker.

This PR will make this easier as such a new format likely won't be able to use an allocator shim at all and this PR is the only way I can think of to remove the need for an allocator shim.

**Comment by apiraino** [comment]
Hey, checking progress on this PR. Any additional comment from T-lang @joshtriplett w.r.t. this [comment](https://github.com/rust-lang/rust/pull/86844#issuecomment-1291113566)? thanks

**Comment by bjorn3** [comment]
By the way https://github.com/rust-lang/rust/issues/104707 might necessitate versioning all internal symbols, including those of the allocator shim, to prevent UB when using different versions of rustc together. Versioning them without this PR will break rust-for-linux as the symbol names would no longer be predictable.

**Comment by apiraino** [comment]
Friendly ping and checking status: was there some progress from T-lang w.r.t. to this PR (cc: @joshtriplett )? Thanks!

**Comment by jackh726** [comment]
@bjorn3 there is no insta-stable behavior here anymore, right? If so, can you update the OP to make that clear?

If this is all only unstable, I think we should get this landed.

**Comment by bjorn3** [comment]
Indeed. You need to define `__rust_no_alloc_shim_is_unstable` to acknowledge that you know it is unstable for linking to succeed. Will update the PR description.

**Comment by bjorn3** [comment]
Rebased

**Comment by pnkfelix** [comment]
I got a chance to play with this a little bit.

----

I was able to demonstrate linking an allocating Rust `#[no_std]` library in with a C driver file, *if* I used a special build of `liballoc` with `--cfg no_global_oom_handling`.

I have not yet put serious effort into compiling a Rust library that also uses `libstd` into a `.o` file that is subsequently linked into a C driver file. The PR description implies this use case is supported; if that is indeed the case, it might be nice to have documentation of how to do it. But I'm not going to block landing this PR on that.

----

As promised, the code had to define `#[no_mangle] static __rust_no_alloc_shim_is_unstable: u8 = 0;` in order to get the build to go through. In the absence of that definition (and that definition *alone*), the error message I got was not quite as good as what I would hope for:

```
/usr/bin/ld: ./librust_obj.a(librust_obj.4yfda7d9s2z2c0cb.rcgu.o): in function `__rust_alloc_error_handler':
4yfda7d9s2z2c0cb:(.text.__rust_alloc_error_handler+0x1): undefined reference to `__rdl_oom'
/usr/bin/ld: c_driver: hidden symbol `__rdl_oom' isn't defined
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
```

in particular, I was *expecting* something in the linkage error about the `__rust_no_alloc_shim_is_unstable` symbol being missing. But I'm okay with this subpar experience here, especially since I don't personally understand why the error message is *this bad* here.

----

@bors r+ rollup=never

**Comment by bjorn3** [comment]
Will rebase tomorrow.

**Comment by bjorn3** [comment]
Rebased

**Comment by bjorn3** [comment]
Fixed miri

File: src/tools/miri/tests/pass/shims/fs.rs [hunk 5]
```diff
@@ -32,14 +32,18 @@ fn main() {
...
     let mut out = Vec::with_capacity(1024);
 
     unsafe {
         extern "Rust" {
-            fn miri_host_to_target_path(path: *const i8, out: *mut i8, out_size: usize) -> usize;
+            fn miri_host_to_target_path(
+                path: *const c_char,
+                out: *mut c_char,
+                out_size: usize,
+            ) -> usize;
```
Comment (bjorn3) [inline src/tools/miri/tests/pass/shims/fs.rs:46]: AArch64 uses unsigned char while x86 uses signed char. Using c_char works on both.

File: src/tools/miri/src/shims/foreign_items.rs [hunk 6]
```diff
@@ -346,11 +345,8 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
 
         match allocator_kind {
             AllocatorKind::Global => {
-                let (body, instance) = this
-                    .lookup_exported_symbol(symbol)?
-                    .expect("symbol should be present if there is a global allocator");
-
-                Ok(EmulateByNameResult::MirBody(body, instance))
+                // `__rust_*` is defined by `#[global_allocator]` if `#[global_allocator]` is used
+                return Ok(EmulateByNameResult::NotSupported);
```
Comment (RalfJung) [inline src/tools/miri/src/shims/foreign_items.rs:349]: I don't understand this comment. It seems unrelated both to the preceding match and the following Unsupported error. What is going on here? The entire set of Miri allocator changes is somewhat mysterious.

File: src/tools/miri/tests/pass/shims/fs.rs (hunk 5 above)
Comment (RalfJung) [reply src/tools/miri/tests/pass/shims/fs.rs:46]: Ah, I've been looking at https://doc.rust-lang.org/nightly/alloc/ffi/struct.CString.html#method.as_ptr which claims the return type is an `i8` pointer but looking at the source shows that rustdoc is omitting some crucial information here.

File: src/tools/miri/src/shims/foreign_items.rs (hunk 6 above)
Comment (bjorn3) [reply src/tools/miri/src/shims/foreign_items.rs:349]: Returning `NotSupported` will be interpreted as there existing no shim. If there is no shim it will fall back to looking for an `#[no_mangle]` symbol with the right name. Exactly what is necessary here. Only when `#[global_allocator]` is not used, should `__rust_*` be redirected to the default `__rdl_*` implementations in libstd.

Comment (RalfJung) [reply src/tools/miri/src/shims/foreign_items.rs:349]: Oh you want to hit the fallback path in emulate_foreign_item. This needs a *ton* of comments explaining what is going on.^^ Both here and also below where you moved the `check_shim` calls around.

Comment (bjorn3) [reply src/tools/miri/src/shims/foreign_items.rs]: This better?

File: src/tools/miri/src/shims/foreign_items.rs [hunk 7]
```diff
@@ -346,11 +345,12 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
...
-                let (body, instance) = this
-                    .lookup_exported_symbol(symbol)?
-                    .expect("symbol should be present if there is a global allocator");
-
-                Ok(EmulateByNameResult::MirBody(body, instance))
+                // When `#[global_allocator]` is used, `__rust_*` is defined by the macro expansion
+                // of this attribute rather than generated by the allocator shim. As such we have
+                // to call the definition produced by `#[global_allocator]` instead of the shim like
+                // in the case of `#[global_allocator]` not existing. Somewhat unintuitively doing
+                // so is done by returning `NotSupported`.
```
Comment (RalfJung) [inline src/tools/miri/src/shims/foreign_items.rs:352]:
```suggestion
                // When `#[global_allocator]` is used, `__rust_*` is defined by the macro expansion
                // of this attribute. As such we have to call an exported Rust function,
                // and not execute any Miri shim. Somewhat unintuitively doing so is done
                // by returning `NotSupported`, which triggers the `lookup_exported_symbol`
                // fallback case in `emulate_foreign_item`.
```

File: src/tools/miri/src/shims/foreign_items.rs [hunk 8]
```diff
@@ -535,11 +535,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
...
             // Rust allocation
             "__rust_alloc" | "miri_alloc" => {
-                let [size, align] = this.check_shim(abi, Abi::Rust, link_name, args)?;
-                let size = this.read_machine_usize(size)?;
-                let align = this.read_machine_usize(align)?;
-
                 let default = |this: &mut MiriInterpCx<'mir, 'tcx>| {
+                    // Only call `check_shim` when `#[global_allocator]` isn't used. The macro
+                    // expansion of `#[global_allocator]` defines this symbol and `check_shim`
+                    // checks that there exists no definition of a shim.
```
Comment (RalfJung) [inline src/tools/miri/src/shims/foreign_items.rs:541]:
```suggestion
                    // Only call `check_shim` when `#[global_allocator]` isn't used. When that
                    // macro is used, we act like no shim exists, so that the exported function can run.
```

**Comment by RalfJung** [comment]
Miri changes LGTM.

**Comment by workingjubilee** [comment]
Alas, it seems this needs another rebase.

**Comment by bjorn3** [comment]
Done. Thanks for the heads up!

**Comment by bjorn3** [comment]
``

**Comment by bjorn3** [comment]
Bors forgot about this PR.

@bors r=pnkfelix rollup=never

**Comment by bjorn3** [comment]
Rebased to fix the merge conflict and fixed the test (previously I needed to update a codegen test, but this no longer seems to be necessary, so I dropped the commit).

**Comment by bjorn3** [comment]
For some reason it seems the codegen test starts at vtable.2 on some targets and vtable.0 on other targets. How should I handle this?

**Comment by bjorn3** [comment]
I adapted the test checks such that the test should pass in both cases.

File: tests/codegen/debug-vtable.rs [hunk 9]
```diff
@@ -11,8 +11,6 @@
 
 // Make sure that vtables don't have the unnamed_addr attribute when debuginfo is enabled.
 // This helps debuggers more reliably map from dyn pointer to concrete type.
-// CHECK: @vtable.0 = private constant <{
-// CHECK: @vtable.1 = private constant <{
```
Comment (pnkfelix) [inline tests/codegen/debug-vtable.rs:15]: I don't know whether to be delighted or horrified that this "fix" works to handle both cases here. :)

**Comment by bjorn3** [comment]
I think I fixed the test that failed on MSVC.

**Comment by RalfJung** [comment]
@pnkfelix do you want to take a look at the adjusted test or should we just send it to CI again? I can't tell if that test change is harmless or not.

**Comment by pnkfelix** [comment]
@RalfJung weird, I thought I had posted approval on this. Maybe I'm thinking of a different PR.

**Comment by pnkfelix** [comment]
oh I see, some test had to be changed again

**Comment by pnkfelix** [comment]
@bors r+ rollup=never

(The test changes is relatively harmless. I am wondering whether there can be unexpected divergences in behavior between Linux and MSVC due to the two command lines having further divrgence (with respect to things like the ordering of the input object files and rlib files), but I'm not going to worry about that for this test at this point.)

**Comment by apiraino** [comment]
@bjorn3 I think this just needs a rebase and then we're good to go? thanks!

@rustbot author

**Comment by bjorn3** [comment]
I was waiting on if disabling the failing test for msvc in https://github.com/rust-lang/rust/pull/106560 would be accepted. If so I will rebase this PR and disable failing test for msvc of this PR too.

**Comment by bjorn3** [comment]
Rebased and just like with https://github.com/rust-lang/rust/pull/106560 I disabled the test on MSVC for now.

File: library/alloc/src/alloc.rs [hunk 10]
```diff
@@ -90,7 +93,14 @@ pub use std::alloc::Global;
 #[must_use = "losing the pointer will leak memory"]
 #[inline]
 pub unsafe fn alloc(layout: Layout) -> *mut u8 {
-    unsafe { __rust_alloc(layout.size(), layout.align()) }
+    unsafe {
+        // Make sure we don't accidentally allow omitting the allocator shim in
+        // stable code until it is actually stabilized.
+        #[cfg(not(bootstrap))]
+        core::ptr::read_volatile(&__rust_no_alloc_shim_is_unstable);
```
Comment (bjorn3) [inline library/alloc/src/alloc.rs:100]: I did expect the perf regression to come from this change. It is a single extra instruction on the allocation path to ensure `__rust_no_alloc_shim_is_unstable` must be defined if no allocator shim is linked in. The only way I can think of that guarantees that it isn't possible to link without defining `__rust_no_alloc_shim_is_unstable` would be to put `__rust_alloc` and an item referencing `__rust_no_alloc_shim_is_unstable` in the same COMDAT group, but this isn't possible for all object file formats and rust doesn't have a way to do this without `global_asm!()`.

**Comment by pnkfelix** [comment]
@bjorn3 when you say:

> I did expect the perf regression to come from this change.

are you referring to the increase in binary object file size?

**Comment by bjorn3** [comment]
I'm referring to the instruction count regression. I hadn't noticed the binary object file size regression.

**Comment by pnkfelix** [comment]
* The 0.2% hit to primary benchmark serde_derive check-incr_unchanged is easily justified by the feature addition here.
* The more interesting question is 44 primary benchmarks saw a regression to their binary size. However, the only one of those of note, in my opinion, is ripgrep, which suffered a 0.43% increase to binary size on various opt scenarios.
* marking as triaged.

@rustbot label: perf-regression-triaged

File: library/alloc/src/alloc.rs (hunk 10 above)
Comment (RReverser) [reply library/alloc/src/alloc.rs:100]: It's unfortunate that this symbol is now present in every single allocation path, especially where it enlarges the output binary for platforms like Wasm.

Comment (bjorn3) [reply library/alloc/src/alloc.rs:100]: That would prevent LLVM from optimizing allocations away I think as `__rust_no_alloc_shim_is_unstable` is not considered to be an allocator function by LLVM, while `__rust_alloc` is. That said, I just opened a draft PR which will enable doing away with the allocator shim entirely in the future, after which I hope it will be much easier to request stabilization of support for not using the allocator shim and thus remove this symbol entirely.
